### ISSUE TITLE
Rework the Auth0 Guide

### ISSUE THREAD

**Issue by zachdaniel** [team-alembic/ash_authentication#443]
Based on feedback from multiple users, for example: https://elixirforum.com/t/onboarding-feedback-the-good-the-bad-and-the-ugly/58527/6?u=zachdaniel

Ultimately what should be done is that we should rework the guide to assume that the user has not already set up `ash_authentication`, and then highlight some areas that might be different if they already have set it up for password authentication, for example.

**Comment by dewetblomerus** [comment]
I am running into what feels like a potential dead-end after following the current guide.

I do not think this is an actual issue with the code, just with the guide, so I'm not opening a separate issue.

After following the guide, my app is vulnerable to account takeover, here are the steps to reproduce:

1. Enable Gmail and Email/Password connectors in Auth0
2. Sign up with Gmail, now my app considers the user to be dewetblomerus@gmail.com
3. In an incognito tab, sign up with Username/Password and choose the same dewetblomerus@gmail.com
4. Do not verify the email address
5. AshAuthentication considers me to be the same user as the one from Gmail, even though the accounts were not linked in Auth0.

### What I think we might need to do

The `user_info` that comes from Auth0 includes a `sub`, these are not the same for the two users.

`"sub" => "google-oauth2|redacted-numbers"`
`"sub" => "auth0|redacted-numbers-and-letters"`

If instead of:

```
  identities do
    identity :unique_email, [:email]
  end
```

We made the `identity :unique_auth_id, [:auth_id]` or maybe just `identity :unique_sub, [:sub]`

Then the different Auth0 accounts would be considered separate.

### What I do not know but I fell about 83% sure

[If they are linked](https://auth0.com/docs/manage-users/user-accounts/user-account-linking), will the sub change? I think they will not. So if an Ash user wants to build user linking after following a guide that caused the Ash user to be correlated to the Auth0 user by `sub`, they will be able to do that. I would love for someone with more Auth0 experience to weigh in.

**Comment by jimsynz** [comment]
Yeah, this seems like an oversight on my part. We do have the `UserIdentity` extension and resource which we can use to track a user's multiple identities, but it all still comes down to whatever we decide to use for the unique identity for the user.  At the very least I think we should update the guides to explain how this could happen in the case where there are multiple strategies enabled.

**Comment by dewetblomerus** [comment]
No worries, it already felt like magic when I followed a guide, got auth working, and I have the absolute minimum amount of code to maintain. The `UserIdentity` extension looks cool.

One more reason to use the `auth0_id` is that if someone changes their email address, using the email as unique identity would mean that we need to change the email in our database and on Auth0 at the same time, if the email was only changed in Auth0, we would end up creating a new user. If we use the `auth0_id`, the email just needs to be changed in Auth0 and then it automatically gets changed in our app on the next login.


What do you think of a PR to the guide that changes the action to:

```
    create :register_with_auth0 do
      argument :user_info, :map, allow_nil?: false
      argument :oauth_tokens, :map, allow_nil?: false
      upsert? true
      upsert_identity :unique_auth0_id

      # Required if you have token generation enabled.
      change AshAuthentication.GenerateTokenChange

      # Required if you have the `identity_resource` configuration enabled.
      change AshAuthentication.Strategy.OAuth2.IdentityChange

      change fn changeset, _ ->
        user_info = Ash.Changeset.get_argument(changeset, :user_info)

        changes = %{
          "email" => Map.get(user_info, "email"),
          "auth0_id" => Map.get(user_info, "sub")
        }

        Ash.Changeset.change_attributes(
          changeset,
          changes
        )
      end
```

I would also wan to incorporate whatever I learn from [here](https://elixirforum.com/t/do-not-touch-updated-at-when-upsert-has-no-effect/59459)

**Comment by jimsynz** [comment]
Absolutely. Rewrite the whole thing if you want 😂

**Comment by jimsynz** [comment]
@dewetblomerus were you still intending to work up a PR for this or should I add it to my backlog?
